### PR TITLE
[ntuple] Add RRVecField for type-erased I/O of RVecs

### LINKED ISSUE #6347: [ntuple] Off-by-one-byte error when deserializing arrays as RVecs
Reading `RVecs` with `RNTuple` incurs in an off-by-one-byte error, because all `RVec<T>` are deserialized as `std::vector<char>`, but the position of the data buffer is different for `RVecs`:

```
>>> p ((ROOT::VecOps::RVec<int>*)0x5555596c72a0)->data()
$983 = (int *) 0x5555581b2918
>>> p ((std::vector<char, std::allocator<char> > *)0x5555596c72a0)->data()
$984 = 0x5555581b2910
```

I think the problematic logic is [here](https://github.com/root-project/root/blob/8d5b55f51ad8449d83fc0db8f49217e556c34036/tree/ntuple/v7/src/RField.cxx#L660).

A second issue related to RVecs is that at RField.cxx:638 we use RVectorField for reading RVecs, but this means that the parent RField will have type name `std::vector<...>` even if it should (probably?) be `RVec<...>`.

#6336 contains a reproducer.

### PR DESCRIPTION
**Pull request by eguiraud**
This PR:
- adds the RRVecField type for type-erased I/O of RVecs with RNTuple
- removes some special-casing of `RVec<bool>` which is unnecessary now that we have RVec 2.0
- adds RPrintValueVisitor support for RVecs
- adds tests for type-erased I/O of RVecs
- adds tests for `ntuple->Show` + RVecs
- adds tests for interop I/O of std::vectors and RVecs (i.e. writing one and reading the other)

## Checklist:

- [x] tested changes locally
- [x] updated the docs (if necessary): I don't think it's necessary

This PR fixes #6347 .

### CODE REVIEW

File: tree/ntuple/v7/src/RField.cxx [hunk 1]
```diff
@@ -1083,6 +1083,211 @@ void ROOT::Experimental::RVectorField::AcceptVisitor(Detail::RFieldVisitor &visi
...
+   // padding might be present:
+   // - between fCapacity and the char[] buffer aligned like T
+   // - after the char[] buffer
+
+   const auto dataMemberSz = sizeof(void *) + 2 * sizeof(int32_t);
+   const auto alignOfT = fSubFields[0]->GetAlignment();
+   const auto sizeOfT = fSubFields[0]->GetValueSize();
+
+   // mimic the logic of RVecInlineStorageSize, but at runtime
+   const auto inlineStorageSz = [&] {
```
Comment (jblomer) [inline tree/ntuple/v7/src/RField.cxx:1243]: Not sure why we need a lambda here.

File: tree/ntuple/v7/src/RField.cxx [hunk 2]
```diff
@@ -1083,6 +1083,211 @@ void ROOT::Experimental::RVectorField::AcceptVisitor(Detail::RFieldVisitor &visi
...
+
+   std::vector<Detail::RFieldValue> result;
+   for (int32_t i = 0; i < size; ++i) {
+      auto elementValue = fSubFields[0]->CaptureValue(reinterpret_cast<void *>(std::intptr_t(begin) + i * fItemSize));
+      result.emplace_back(std::move(elementValue));
+   }
+   return result;
+}
+
+size_t ROOT::Experimental::RRVecField::GetValueSize() const
```
Comment (jblomer) [inline tree/ntuple/v7/src/RField.cxx:1224]: It might make sense to cache the size.

File: tree/ntuple/v7/src/RField.cxx [hunk 3]
```diff
@@ -1083,6 +1083,211 @@ void ROOT::Experimental::RVectorField::AcceptVisitor(Detail::RFieldVisitor &visi
...
+}
+
+std::vector<ROOT::Experimental::Detail::RFieldValue>
+ROOT::Experimental::RRVecField::SplitValue(const Detail::RFieldValue &value) const
+{
+   void **rvecAddr = reinterpret_cast<void**>(value.GetRawPtr());
+   // void* fBegin is the first data member
+   void *begin = *rvecAddr;
+   // int32_t fSize is the second data member (after 1 void*)
+   const int32_t size = *reinterpret_cast<int32_t *>(rvecAddr + 1);
```
Comment (jblomer) [inline tree/ntuple/v7/src/RField.cxx:1213]:
```suggestion
   const std::int32_t size = *reinterpret_cast<std::int32_t *>(rvecAddr + 1);
```

File: tree/ntuple/v7/src/RField.cxx [hunk 4]
```diff
@@ -1083,6 +1083,211 @@ void ROOT::Experimental::RVectorField::AcceptVisitor(Detail::RFieldVisitor &visi
...
+   R__ASSERT(size >= 0);
+
+   std::vector<Detail::RFieldValue> result;
+   for (int32_t i = 0; i < size; ++i) {
```
Comment (jblomer) [inline tree/ntuple/v7/src/RField.cxx:1217]:
```suggestion
   for (std::int32_t i = 0; i < size; ++i) {
```

File: tree/ntuple/v7/src/RField.cxx [hunk 5]
```diff
@@ -1083,6 +1083,211 @@ void ROOT::Experimental::RVectorField::AcceptVisitor(Detail::RFieldVisitor &visi
...
+   // - void *fBegin
+   // - int32_t fSize
+   // - int32_t fCapacity
+   // - the char[] inline storage, which is aligned like T
+   //
+   // padding might be present:
+   // - between fCapacity and the char[] buffer aligned like T
+   // - after the char[] buffer
+
+   const auto dataMemberSz = sizeof(void *) + 2 * sizeof(int32_t);
```
Comment (jblomer) [inline tree/ntuple/v7/src/RField.cxx:1238]:
```suggestion
   const auto dataMemberSz = sizeof(void *) + 2 * sizeof(std::int32_t);
```

File: tree/ntuple/v7/src/RField.cxx [hunk 6]
```diff
@@ -1083,6 +1083,211 @@ void ROOT::Experimental::RVectorField::AcceptVisitor(Detail::RFieldVisitor &visi
...
+   return Detail::RFieldValue(/*captureTag*/true, this, where);
+}
+
+void ROOT::Experimental::RRVecField::DestroyValue(const Detail::RFieldValue& value, bool dtorOnly)
+{
+   void **rvecAddr = reinterpret_cast<void**>(value.GetRawPtr());
+   // void* fBegin is the first data member
+   void *begin = *rvecAddr;
+   // int32_t fSize is the second data member (after 1 void*, no padding expected)
+   const int32_t* sizePtr = reinterpret_cast<int32_t *>(rvecAddr + 1);
```
Comment (jblomer) [inline tree/ntuple/v7/src/RField.cxx:1176]:
```suggestion
   const std::int32_t* sizePtr = reinterpret_cast<std::int32_t *>(rvecAddr + 1);
```

File: tree/ntuple/v7/src/RField.cxx [hunk 7]
```diff
@@ -1083,6 +1083,211 @@ void ROOT::Experimental::RVectorField::AcceptVisitor(Detail::RFieldVisitor &visi
...
+   const int32_t size = *sizePtr;
```
Comment (jblomer) [inline tree/ntuple/v7/src/RField.cxx:1177]:
```suggestion
   const std::int32_t size = *sizePtr;
```

File: tree/ntuple/v7/src/RField.cxx [hunk 8]
```diff
@@ -1083,6 +1083,211 @@ void ROOT::Experimental::RVectorField::AcceptVisitor(Detail::RFieldVisitor &visi
...
+   const auto dataMemberSz = sizeof(void *) + 2 * sizeof(int32_t);
+   const auto alignOfT = fSubFields[0]->GetAlignment();
+   auto paddingMiddle = dataMemberSz % alignOfT;
+   if (paddingMiddle != 0)
+      paddingMiddle = alignOfT - paddingMiddle;
+   const bool isSmall = begin == (rvecAddr + dataMemberSz + paddingMiddle);
+
+   const bool owns = capacity != -1;
+   if (!dtorOnly && !isSmall && owns)
+      free(begin);
```
Comment (jblomer) [inline tree/ntuple/v7/src/RField.cxx:1198]: Are we sure that the memory pointed to by `begin` was allocated by `malloc` (not by `new[]`)? E.g. line 1130 doesn't.

File: tree/ntuple/v7/src/RField.cxx [hunk 9]
```diff
@@ -1083,6 +1083,211 @@ void ROOT::Experimental::RVectorField::AcceptVisitor(Detail::RFieldVisitor &visi
...
+      fSubFields[0]->Append(elementValue);
+   }
+
+   fColumns[0]->Append(Detail::RColumnElement<ClusterSize_t>(&fNWritten));
+   fNWritten += size;
+}
+
+void ROOT::Experimental::RRVecField::ReadGlobalImpl(NTupleSize_t globalIndex, Detail::RFieldValue *value)
+{
+   // TODO as a performance optimization, we could assign values to elements of the inline buffer
```
Comment (jblomer) [inline tree/ntuple/v7/src/RField.cxx:1121]: I think that's an important optimization

File: tree/ntuple/v7/src/RField.cxx [hunk 10]
```diff
@@ -1083,6 +1083,211 @@ void ROOT::Experimental::RVectorField::AcceptVisitor(Detail::RFieldVisitor &visi
...
+   // this might not be the case if T in RVec<T> is over-aligned.
+   auto *buffer = new char[nItems * fItemSize];
+   for (auto i = 0u; i < nItems; ++i) {
+      auto itemValue = fSubFields[0]->GenerateValue(buffer + (i * fItemSize));
+      fSubFields[0]->Read(collectionStart + i, &itemValue);
+   }
+
+   void **rvecAddr = reinterpret_cast<void**>(value->GetRawPtr());
+   // void* fBegin is the first data member
+   *rvecAddr = buffer;
```
Comment (jblomer) [inline tree/ntuple/v7/src/RField.cxx:1138]: That leaks the memory previously allocated for `buffer`. We could first check if the size is large enough and only reallocate if it isn't.  Perhaps reallocating `factor * current size` to limit the number of allocations.

**Review by jblomer — CHANGES_REQUESTED** [review]
Awesome!  Some comments inline. I marked some `std::int32_t` instead of `int32_t` but not all. I think it would be also good to warn in the documentation of `RVec` that changes to the memory allocation and to the class layout need to be reflected here (perhaps that's already there).

File: tree/ntuple/v7/src/RField.cxx (hunk 2 above)
Comment (eguiraud) [reply tree/ntuple/v7/src/RField.cxx:1224]: Should we evaluate it once in the ctor or evaluate it lazily and cache it?

File: tree/ntuple/v7/src/RField.cxx (hunk 1 above)
Comment (eguiraud) [reply tree/ntuple/v7/src/RField.cxx:1243]: It's the IIFE pattern, e.g. https://www.bfilipek.com/2016/11/iife-for-complex-initialization.html?m=1

It's to clearly distinguish the not-so-trivial logic that's only there to initialize this variable.
I can just put everything inline if you prefer

File: tree/ntuple/v7/src/RField.cxx (hunk 9 above)
Comment (eguiraud) [reply tree/ntuple/v7/src/RField.cxx:1121]: I agree! Do you want it in this PR or can I open a subsequent one with just this optimization?

File: tree/ntuple/v7/inc/ROOT/RFieldVisitor.hxx [hunk 11]
```diff
@@ -64,6 +64,7 @@ public:
    virtual void VisitUInt8Field(const RField<std::uint8_t> &field) { VisitField(field); }
    virtual void VisitVectorField(const RVectorField &field) { VisitField(field); }
    virtual void VisitVectorBoolField(const RField<std::vector<bool>> &field) { VisitField(field); }
+   virtual void VisitRVecField(const RRVecField &field) { VisitField(field); }
```
[the remark this answers is not in the document]
Comment (eguiraud) [reply tree/ntuple/v7/inc/ROOT/RFieldVisitor.hxx:67]: implemented it, wrote a test, then realized that the test was testing `RField<RVec<T>>` and not `RRVecField` and that one also does not implement value printing for `RVecs`. work in progress but now related/depending on https://github.com/root-project/root/pull/8770#discussion_r679305972

File: tree/ntuple/v7/src/RField.cxx [hunk 12]
```diff
@@ -181,13 +183,11 @@ ROOT::Experimental::Detail::RFieldBase::Create(const std::string &fieldName, con
       std::string itemTypeName = normalizedType.substr(12, normalizedType.length() - 13);
       auto itemField = Create(GetNormalizedType(itemTypeName), itemTypeName);
       result = std::make_unique<RVectorField>(fieldName, itemField.Unwrap());
-   } else if (normalizedType == "ROOT::VecOps::RVec<bool>") {
-      result = std::make_unique<RField<ROOT::VecOps::RVec<bool>>>(fieldName);
    } else if (normalizedType.substr(0, 19) == "ROOT::VecOps::RVec<") {
       // For the time being, we silently read RVec fields as std::vector
```
[the remark this answers is not in the document]
Comment (eguiraud) [reply tree/ntuple/v7/src/RField.cxx]: removed

File: tree/ntuple/v7/src/RField.cxx [hunk 13]
```diff
@@ -1083,6 +1083,211 @@ void ROOT::Experimental::RVectorField::AcceptVisitor(Detail::RFieldVisitor &visi
...
+   return std::make_unique<RRVecField>(newName, std::move(newItemField));
+}
+
+void ROOT::Experimental::RRVecField::AppendImpl(const Detail::RFieldValue &value)
+{
+   void **rvecAddr = reinterpret_cast<void**>(value.GetRawPtr());
+   // void* fBegin is the first data member
+   void *begin = *rvecAddr;
+   // int32_t fSize is the second data member (after 1 void*)
+   const int32_t size = *reinterpret_cast<int32_t *>(rvecAddr + 1);
```
[the remark this answers is not in the document]
Comment (eguiraud) [reply tree/ntuple/v7/src/RField.cxx]: should be fixed everywhere

File: tree/ntuple/v7/src/RField.cxx (hunk 8 above)
Comment (eguiraud) [reply tree/ntuple/v7/src/RField.cxx:1197]: ah, you are right. Is `RField::DestroyValue` with `!dtorOnly` always called on memory that `RField` allocated? If yes, I can just make line 1130 a malloc. If not, I'm not sure how to proceed

File: tree/ntuple/v7/inc/ROOT/RField.hxx [hunk 14]
```diff
@@ -385,6 +385,43 @@ public:
...
+   ClusterSize_t fNWritten;
+
+protected:
+   std::unique_ptr<Detail::RFieldBase> CloneImpl(std::string_view newName) const final;
+   void AppendImpl(const Detail::RFieldValue &value) final;
+   void ReadGlobalImpl(NTupleSize_t globalIndex, Detail::RFieldValue *value) final;
+
+public:
+   RRVecField(std::string_view fieldName, std::unique_ptr<Detail::RFieldBase> itemField);
+   RRVecField(RRVecField &&other) = default;
```
[the remark this answers is not in the document]
Comment (eguiraud) [reply tree/ntuple/v7/inc/ROOT/RField.hxx]: done

File: tree/ntuple/v7/src/RField.cxx (hunk 2 above)
Comment (eguiraud) [reply tree/ntuple/v7/src/RField.cxx:1223]: (at construction time)

File: tree/ntuple/v7/src/RField.cxx (hunk 8 above)
Comment (eguiraud) [reply tree/ntuple/v7/src/RField.cxx:1197]: will double-check but should be ok

File: tree/ntuple/v7/src/RField.cxx (hunk 9 above)
Comment (eguiraud) [reply tree/ntuple/v7/src/RField.cxx:1120]: (will add in a later PR)

File: tree/ntuple/v7/src/RField.cxx [hunk 15]
```diff
@@ -1083,6 +1083,211 @@ void ROOT::Experimental::RVectorField::AcceptVisitor(Detail::RFieldVisitor &visi
...
+   // TODO as a performance optimization, we could assign values to elements of the inline buffer
+   // if size < inline buffer size: we save one allocation here and usage of the RVec skips a pointer indirection
+
+   ClusterSize_t nItems;
+   RClusterIndex collectionStart;
+   fPrincipalColumn->GetCollectionInfo(globalIndex, &collectionStart, &nItems);
+
+   // we trust that `new` returns a buffer with large enough alignment.
+   // this might not be the case if T in RVec<T> is over-aligned.
+   auto *buffer = new char[nItems * fItemSize];
```
Comment (eguiraud) [inline tree/ntuple/v7/src/RField.cxx:1129]: this should be a malloc

**Comment by eguiraud** [comment]
@phsft-bot build

**Comment by eguiraud** [comment]
Rebased to fix conflicts due to changes in RNTuple merged in the meanwhile. Still have to address https://github.com/root-project/root/pull/8770#discussion_r731013612 before merging.

**Comment by eguiraud** [comment]
@jblomer about the discussion at https://github.com/root-project/root/pull/8770#discussion_r730729833 , I remember that for RVec we said we would not call destructors for collection elements at every entry, but I see that RVectorField does at line 1039? https://github.com/root-project/root/blob/7e709f8810023ad184747e4c6f250f3617bb25dc/tree/ntuple/v7/src/RField.cxx#L1036-L1044

Also if the `resize` at line 1041 needs to reallocate is it correct to only call `GenerateValue` for new elements? If my understanding is correct, the _bytes_ of the old elements will be correct, but no placement new for a class of the correct type will have been called in the new memory location so _technically_ the object's lifetimes is not started.

(asking about RVectorField to understand what to do with RRVecField)

**Comment by jblomer** [comment]
> @jblomer about the discussion at [#8770 (comment)](https://github.com/root-project/root/pull/8770#discussion_r730729833) , I remember that for RVec we said we would not call destructors for collection elements at every entry, but I see that RVectorField does at line 1039?

It does so only for the left-over elements when the collection shrinks.

> Also if the `resize` at line 1041 needs to reallocate is it correct to only call `GenerateValue` for new elements? If my understanding is correct, the _bytes_ of the old elements will be correct, but no placement new for a value of the correct type will have been called in the new memory location so _technically_ the object's lifetimes is not started.

Good point! I created #10520.

**Comment by eguiraud** [comment]
> It does so only for the left-over elements when the collection shrinks.

uhm why are leftover elements different from elements that are replaced? if some resource management logic has to be called when an element is removed, it might also have to be called when it is replaced?

File: tree/ntuple/v7/inc/ROOT/RNTupleModel.hxx [hunk 16]
```diff
@@ -195,7 +195,7 @@ public:
    REntry *GetDefaultEntry() const;
 
    RFieldZero *GetFieldZero() const { return fFieldZero.get(); }
-   const Detail::RFieldBase *GetField(std::string_view fieldName) const;
+   Detail::RFieldBase *GetField(std::string_view fieldName) const;
```
Comment (jblomer) [inline tree/ntuple/v7/inc/ROOT/RNTupleModel.hxx:198]: As per discussion in #10408, can this go back to `const`?

**Review by jblomer — CHANGES_REQUESTED** [review]
Looks good to me, except one tiny comment on the change in the `RNTupleModel` API

File: tree/ntuple/v7/inc/ROOT/RNTupleModel.hxx (hunk 16 above)
Comment (eguiraud) [reply tree/ntuple/v7/inc/ROOT/RNTupleModel.hxx]: done

**Comment by eguiraud** [comment]
This should be ready to go, with two performance optimizations (marked as TODOs) that are left for future PRs (this one is already terribly large, I think).

**Review by jblomer — APPROVED** [review]
Great! I'm looking forward to adjusting the ntuple data source to this.

**Pull request merged**